folsom/src/main/java/com/spotify/folsom/client/Request.java

```diff
@@ -28,9 +28,11 @@
 
 public abstract class Request<V> extends CompletableFuture<V> {
   protected final byte[] key;
+  protected final long createdNanos;
 
   protected Request(byte[] key) {
     this.key = key;
+    createdNanos = System.nanoTime();
```

Comment:
Thinking more about it - if someone creates a lot of requests in a batch, they will end up being queued, so the creation time of the request is not what we want.

Instead, I think the TimeoutChecker should keep state of when the head-request was first noticed. Which is what the code already does.

I would however be ok with checking for timeouts more frequently with the pollIntervalMillis change.

Reply:
I can see value in the current behavior (check when the request was first noticed).
at the same time -
from client point of view, if I define a timeout of 100ms and send a request which then spends 95 msec in a queue and then 10 msec in flight, I still want it to time out since it took longer than 100 msec overall.
we use tons of microservices and they all have specific SLAs. accumulating extra processing time during a request chain may lead to negative end-user experience. I would rather get a timeout after 100 msec from the moment when I first initialize the Memcached request. this timeout can serve as a signal that I may need to adjust my batch size, maybe prioritize local cache,  or find some other solution -  but it is a better outcome for me than requests taking longer than max configured time of 100 msec.

another option is to separate "max total time" and "max flight time" options, but that may be an overkill because many apps probably won't care about these two, and a few highly loaded apps will *probably* want the total time

Reply:
I think there are two separate use cases here:
1. detecting and recovering from broken connections (what's already in place, although parts of this PR could be used to improve it)
2. setting an upper limit of acceptable wait-time for requests.

Use case 2 could be somewhat implemented externally to folsom by adding timeouts (like this perhaps: http://iteratrlearning.com/java9/2016/09/13/java9-timeouts-completablefutures.html ) though that wouldn't have other nice features such as fast-failing the folsom requests if we're falling behind.

I would prefer to keep the existing functionality, but possibly add new functionality to timeout individual requests